src/ripple/shamap/impl/SHAMapNodeID.cpp

```diff
@@ -49,106 +48,88 @@ SHAMapNodeID::Masks(int depth)
             entry[mask_size - 1] = selector;
         }
     };
+
     static masks_t const masks;
     return masks.entry[depth];
 }
 
 // canonicalize the hash to a node ID for this depth
-SHAMapNodeID::SHAMapNodeID(int depth, uint256 const& hash)
-    : mNodeID(hash), mDepth(depth)
+SHAMapNodeID::SHAMapNodeID(unsigned int depth, uint256 const& hash)
+    : id_(hash), depth_(depth)
 {
-    assert((depth >= 0) && (depth < 65));
-    assert(mNodeID == (mNodeID & Masks(depth)));
-}
-
-SHAMapNodeID::SHAMapNodeID(void const* ptr, int len)
-{
-    if (len < 33)
-        mDepth = -1;
-    else
-    {
-        std::memcpy(mNodeID.begin(), ptr, 32);
-        mDepth = *(static_cast<unsigned char const*>(ptr) + 32);
-    }
-}
-
-std::string
-SHAMapNodeID::getString() const
-{
-    if ((mDepth == 0) && (mNodeID.isZero()))
-        return "NodeID(root)";
-
-    return "NodeID(" + std::to_string(mDepth) + "," + to_string(mNodeID) + ")";
-}
-
-void
-SHAMapNodeID::addIDRaw(Serializer& s) const
-{
-    s.addBitString(mNodeID);
-    s.add8(mDepth);
+    assert(depth <= SHAMap::leafDepth);
+    assert(id_ == (id_ & depthMask(depth)));
 }
 
 std::string
 SHAMapNodeID::getRawString() const
 {
     Serializer s(33);
-    addIDRaw(s);
+    s.addBitString(id_);
+    s.add8(depth_);
     return s.getString();
 }
 
-// This can be optimized to avoid the << if needed
 SHAMapNodeID
-SHAMapNodeID::getChildNodeID(int m) const
+SHAMapNodeID::getChildNodeID(unsigned int m) const
+{
+    assert(m < SHAMap::branchFactor);
+
+    // A SHAMap has exactly 65 levels, so nodes must not exceed that
+    // depth; if they do, this breaks the invariant of never allowing
+    // the construction of a SHAMapNodeID at an invalid depth. We assert
+    // to catch this in debug builds.
+    //
+    // We throw (but never assert) if the node is at level 64, since
+    // entries at that depth are leaf nodes and have no children and even
+    // constructing a child node from them would break the above invariant.
+    assert(depth_ <= SHAMap::leafDepth);
```

Comment:
I did read the comment, but I'm still not following why this is `<=` and not `<`. I'd expect anything we'd throw a `logic_error` for is something we'd also want to `assert` on.

Reply:
You know, I had convinced myself that this was right but now I don't get it either.